### ISSUE TITLE
Cannot train finetuned model: ValueError: Shapes (2575, 128) and (2570, 128) are incompatible

### ISSUE THREAD

**Issue by nataliiao** [RasaHQ/rasa#8496]
**Rasa version**: 2.3.4

**Rasa SDK version** (if used & relevant): Not relevant

**Rasa X version** (if used & relevant): Not relevant

**Python version**:  3.7

**Operating system** (windows, osx, ...): macOS Catalina 10.15.7

**Issue**:  
When I try to finetune the model and use it for further finetuning I get an error:
ValueError: Shapes (2575, 128) and (2570, 128) are incompatible
I configured CountVectorsFeaturizer and RegexFeaturizer as it’s said in the docs, but it didn’t help.

**Error (including full traceback)**:
```
  File "uvloop/loop.pyx", line 1456, in uvloop.loop.Loop.run_until_complete
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/rasa/train.py", line 174, in train_async
    finetuning_epoch_fraction=finetuning_epoch_fraction,
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/rasa/train.py", line 353, in _train_async_internal
    finetuning_epoch_fraction=finetuning_epoch_fraction,
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/rasa/train.py", line 396, in _do_training
    finetuning_epoch_fraction=finetuning_epoch_fraction,
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/rasa/train.py", line 798, in _train_nlu_with_validated_data
    called_from_combined_training=train_path is not None,
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/rasa/train.py", line 872, in _nlu_model_for_finetuning
    finetuning_epoch_fraction=finetuning_epoch_fraction,
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/rasa/nlu/model.py", line 334, in load
    should_finetune=new_config is not None,
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/rasa/nlu/model.py", line 402, in create
    component_meta, model_metadata.model_dir, model_metadata, **context
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/rasa/nlu/components.py", line 818, in load_component
    component_meta, model_dir, model_metadata, cached_component, **context
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/rasa/nlu/registry.py", line 178, in load_component_by_meta
    component_meta, model_dir, metadata, cached_component, **kwargs
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/rasa/nlu/classifiers/diet_classifier.py", line 1036, in load
    finetune_mode=should_finetune,
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/rasa/nlu/classifiers/diet_classifier.py", line 1115, in _load_model
    finetune_mode=finetune_mode,
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/rasa/nlu/classifiers/diet_classifier.py", line 1152, in _load_model_class
    finetune_mode=finetune_mode,
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/rasa/utils/tensorflow/models.py", line 408, in load
    model.load_weights(model_file_name)
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py", line 2182, in load_weights
    status = self._trackable_saver.restore(filepath, options)
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/tensorflow/python/training/tracking/util.py", line 1320, in restore
    checkpoint=checkpoint, proto_id=0).restore(self._graph_view.root)
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/tensorflow/python/training/tracking/base.py", line 209, in restore
    restore_ops = trackable._restore_from_checkpoint_position(self)  # pylint: disable=protected-access
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/tensorflow/python/training/tracking/base.py", line 908, in _restore_from_checkpoint_position
    visit_queue=visit_queue))
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/tensorflow/python/training/tracking/base.py", line 943, in _single_restoration_from_checkpoint_position
    if child_position.bind_object(trackable=local_object):
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/tensorflow/python/training/tracking/base.py", line 258, in bind_object
    slot_name=slot_restoration.slot_name)
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/tensorflow/python/keras/optimizer_v2/optimizer_v2.py", line 1236, in _create_or_restore_slot_variable
    slot_variable_position.restore(slot_variable)
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/tensorflow/python/training/tracking/base.py", line 209, in restore
    restore_ops = trackable._restore_from_checkpoint_position(self)  # pylint: disable=protected-access
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/tensorflow/python/training/tracking/base.py", line 914, in _restore_from_checkpoint_position
    tensor_saveables, python_saveables))
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/tensorflow/python/training/tracking/util.py", line 297, in restore_saveables
    validated_saveables).restore(self.save_path_tensor, self.options)
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/tensorflow/python/training/saving/functional_saver.py", line 340, in restore
    restore_ops = restore_fn()
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/tensorflow/python/training/saving/functional_saver.py", line 316, in restore_fn
    restore_ops.update(saver.restore(file_prefix, options))
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/tensorflow/python/training/saving/functional_saver.py", line 111, in restore
    restored_tensors, restored_shapes=None)
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/tensorflow/python/training/saving/saveable_object_util.py", line 127, in restore
    self.handle_op, self._var_shape, restored_tensor)
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/tensorflow/python/ops/resource_variable_ops.py", line 311, in shape_safe_assign_variable_handle
    shape.assert_is_compatible_with(value_tensor.shape)
  File "/Users/nataliiao/PycharmProjects/test-rasa-finetune/venv/lib/python3.7/site-packages/tensorflow/python/framework/tensor_shape.py", line 1134, in assert_is_compatible_with
    raise ValueError("Shapes %s and %s are incompatible" % (self, other))
ValueError: Shapes (11587, 128) and (11584, 128) are incompatible
```

**Command or request that led to error**:
```
rasa train -vv --finetune --epoch-fraction 0.5
```

**Content of configuration file (config.yml)** (if relevant):
```yml
language: en
pipeline:
- name: WhitespaceTokenizer
  intent_tokenization_flag: true
  intent_split_symbol: +
- name: RegexFeaturizer
  number_additional_patterns: 100
- name: DucklingEntityExtractor
  url: http://eva-duckling-eva-duckling-svc-headless:80
  timezone: Asia/Jerusalem
- name: EntitySynonymMapper
- name: LanguageModelFeaturizer
  model_weights: "bert-base-uncased"
  model_name: "bert"

- name: CountVectorsFeaturizer
  analyzer: char_wb
  min_ngram: 1
  max_ngram: 4
  additional_vocabulary_size:
    text: 10000
    response: 10000
    action_text: 10000
- name: LexicalSyntacticFeaturizer
  features:
  - - low
    - title
    - upper
  - - BOS
    - EOS
    - low
    - prefix5
    - prefix2
    - suffix5
    - suffix3
    - suffix2
    - upper
    - title
    - digit
  - - low
    - title
    - upper
- name: DIETClassifier
  epochs: 20
  number_of_transformer_layers: 2
  transformer_size: 256
  use_masked_language_model: true
  drop_rate: 0.2
  weight_sparsity: 0.7
  batch_size: [64, 256]
  embedding_dimension: 30
  hidden_layers_sizes:
    text: [512, 128]
  random_seed: 1
  intent_classification: true
  entity_recognition: true
  BILOU_flag: true
  constrain_similarities: true
  model_confidence: linear_norm
- name: FallbackClassifier
  threshold: 0.2
  ambiguity_threshold: 0.1

policies:
- name: AugmentedMemoizationPolicy
  max_history: 10
- name: RulePolicy
  core_fallback_threshold: 0.3
  core_fallback_action_name: action_default_fallback
- name: TEDPolicy
  max_history: 10
  constrain_similarities: true
  model_confidence: linear_norm
  epochs: 20
  batch_size:
  - 32
  - 64
```

**Content of domain file (domain.yml)** (if relevant):
I guess it's not relevant, but can provide it if needed

**Definition of done**
- [x] Determine if this bug is reproducible
- [x] Investigate cause of the bug
- [x] Create an issue for the fix

**Comment by sara-tagger** [comment]
Thanks for raising this issue, @joejuzl will get back to you about it soon✨

###### Please also check out the [docs](https://rasa.com/docs/) and the [forum](https://forum.rasa.com/) in case your issue was raised there too 🤗

**Comment by tttthomasssss** [comment]
I could reproduce the bug with a fresh rasa `2.7.0` installation, trained (and subsequently finetuned) on data from carbon-bot (`phoenix-bot` branch). The problem seems to come from rasa core finetuning - rasa nlu finetuning doesn't appear to be affected.

**Comment by tttthomasssss** [comment]
@aeshky assigned as reviewer

**Comment by tttthomasssss** [comment]
Issue for fixing this problem is #8895.

**Comment by tttthomasssss** [comment]
@nataliiao thanks very much for reporting this! Please find a brief update below:

* A fix for this issue is currently blocked by #8232 and the problem will be fixed when #8232 is merged. 

* The underlying problem is that `LexicalSyntacticFeaturizer` doesn't support incremental training yet, so any config that has this featurizer will run into this error when finetuning.

**Comment by nataliiao** [comment]
@tttthomasssss Thank you!
